packages/react/src/NavList/NavList.tsx

```diff
@@ -275,6 +275,52 @@ const Group: React.FC<NavListGroupProps> = ({title, children, sx: sxProp = defau
   )
 }
 
+export type NavListExpandProps = {
+  children: React.ReactNode
+  label: string
+} & SxProp
+
+const Expand = React.forwardRef<HTMLButtonElement, NavListExpandProps>(({label, children, ...props}, forwardedRef) => {
+  const [expanded, setExpanded] = React.useState(false)
+  const triggerContainer = React.useRef<HTMLLIElement>(null)
+  const [focusTargetParent, setFocusTargetParent] = React.useState<HTMLLIElement>()
+
+  const expandList = () => {
+    // Can only be "true", as trigger is removed after expansion
+    const parentItem = triggerContainer.current?.parentElement as HTMLLIElement
+    setFocusTargetParent(parentItem)
+    setExpanded(true)
+  }
+
+  React.useEffect(() => {
+    if (expanded && focusTargetParent) {
+      // Focus the last 'data-target-focus' element that is a child of current list
+      const focusTargets = Array.from(
+        focusTargetParent.querySelectorAll('[data-target--expand-focus]'),
+      ) as HTMLAnchorElement[]
+
+      focusTargets[focusTargets.length - 1]?.focus()
+    }
+  }, [expanded, focusTargetParent])
+
+  return !expanded ? (
+    <Box as="li" sx={{listStyle: 'none'}} ref={triggerContainer}>
+      <ActionList.Item as="button" aria-expanded="false" ref={forwardedRef} onClick={expandList} {...props}>
+        {label}
+        <ActionList.TrailingVisual>
+          <PlusIcon />
+        </ActionList.TrailingVisual>
+      </ActionList.Item>
+    </Box>
+  ) : (
+    React.Children.map(children, (child, index) => {
+      const isFirstChild = index === 0 ? {'data-target--expand-focus': true} : null
```

Comment:
I'm curious if there's a better way to go about this? The alternative is to utilize a `ref` and pass that to the cloned child - but we'd need to ensure any `ref` passed by the consumer takes priority, while storing that reference somewhere 🤔 

The alternative is the current solution, which is utilizing the `ref` on the expand button to target the container, as `children` cannot have a wrapper (e.g. `<Box ref={ref}>{children}</Box>`). We're essentially adding a `data-*` attribute to the first child cloned, and focusing that when `expanded === true`. 

Would love to hear other ideas, as I am probably missing a few! :eyes:

Reply:
There are two goals for me that come up for this kind of thing:

- Goal: colocate the effect (focus management) with the user interaction (onClick in this case)
- Sub-goal: avoid use of React.Children to promote composition 

The sub-goal isn't necessarily deal breaking since we already do this in the library, but if we could find ways to avoid this technique it will help us out in the long run. The core limitation of this technique is that it forces strict Parent > Child relationships and if people wrap components this composition will break the intended functionality.

First off, curious to hear what you think of these goals and if they align or not 👀 

With these in mind, a couple of approaches came to mind:

- useEffect that runs based on `expanded` (potential for focus-stealing behavior if we're not careful)
- Focus in `onClick` using `flushSync()`
- Focus in `onClick` using a helper that runs a callback in an effect
- Mutation observer on the outer list (probably not ideal since the code for focus is very removed from the actual item onClick)

The hope is that we could write something like:

```tsx
<button
  type="button"
  onClick={() => {
    setExpanded(true);

    // focus item
  }}
/>
```

Probably the biggest challenge with this is figuring out _which items are focus candidates_ since we don't have a common ancestor across elements.

One idea for this that avoids `React.Children` is to use a context provider at the `ShowMoreItems` level. For example:

```tsx
const ShowMoreItemContext = React.createContext<{
  groupId: string;
} | null>(null);
```

In each item, we could then add a data attribute, like in:

```tsx
function Item({ children }: React.PropsWithChildren<{}>) {
  const groupContext = React.useContext(ShowMoreItemContext);
  return (
    <li data-show-more-group={groupContext ? groupContext.groupId : undefined}>
      <button type="button">{children}</button>
    </li>
  );
}
```

Then in our handler we could ``querySelector(`[data-show-more-group="${groupId}"`)`` to get the first element in our group to focus.

```tsx
<button
  type="button"
  onClick={() => {
    setExpanded(true);
    const item = document.querySelector(`[data-show-more-group="${groupId}"`);
    item?.focus();
  }}
/>
```

But the challenge with this is that we have not rendered yet so these elements aren't in the document 😅 This is where things like `flushSync()` could come into play:

```tsx
<button
  type="button"
  onClick={() => {
    flushSync(() => {
      setExpanded(true);
    });
    const item = document.querySelector(`[data-show-more-group="${groupId}"`);
    item?.focus();
  }}
/>
```

Or the effect handler:

```tsx
React.useEffect(() => {
  if (expanded) {
    const item = document.querySelector(`[data-show-more-group="${groupId}"`);
    item?.focus();
  }
}, [expanded, groupId]);
```

_Note: this has a chance to steal focus accidentally due to it being re-run due to dependencies. It'd be great if we could fire this as a one-off and guard against it in the future_

Or some helper that allows us to run code with effects inline when the effects of a component run:

```tsx
<button
  type="button"
  onClick={() => {
    setExpanded(true);
    focus(() => {
      return document.querySelector(`[data-show-more-group="${groupId}"`);
    });
  }}
/>
```

---

Sorry that this ended up being so long 😂 You gave me a lot to chew on. I'd love to hear what you think and learn more about what you've tried out / what would work best here from your perspective 👀

Reply:
This is great stuff, thank you! 

I updated my PR to utilize context instead! I think this is a much better way to handle things, so thank you for the recommendation! I'm curious about the different routes we might have when it comes to focus management. I'm still utilizing `useEffect` to handle things, but I agree that there's a possibility that focus might be set unexpectedly based on certain conditions.

I'm wondering if utilizing `useRef` could be a good way to guard against dependencies causing `.focus()` being called again unexpectedly. The ref would only store `true` or `false` based on if the target has been focused or not, with the permanent value being `true` once focused. 

[I created a quick example of utilizing `useRef`](https://github.com/primer/react/pull/4686/commits/3c0c94259decd22eba88563430eae274031c46e8), and I'm wondering if there's still potential risk here 🤔

Reply:
@TylerJDev oh 100%, I think the `useRef` technique would work great here 👍 

I also opened up: https://github.com/primer/react/pull/4715 as a proposal for extending `useFocus()` to support this use-case too that I'd love to get your thoughts on.